File: examples/c/subscribe_async/subscribe_async.c
/*
 * Copyright 2007-2022 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file subscribe_async.c
 *
 * @brief This example program provides a basic demonstration of subscribing
 * to receive a message from the GMSEC Bus using an asynchronous callback.
 */

#include "../utility.h"

#include <gmsec5_c.h>

#include <stdlib.h>


// Define custom callback to receive messages
//
GMSEC_BOOL g_gotMsg = GMSEC_FALSE;

void myCallback(GMSEC_Connection conn, GMSEC_Message msg)
{
	g_gotMsg = GMSEC_TRUE;

	GMSEC_INFO("myCallback:\n%s", messageToXML(msg, NULL));

	// Acknowledge processing the message
	messageAcknowledge(msg, NULL);
}

GMSEC_BOOL gotMessage()
{
	return g_gotMsg;
}


// Define custom callback to receive events
//
// Although not required, an event callback can assist with determining
// whether we have a connection to the GMSEC Bus, or in the case of this
// example program, whether the Auto-Dispatcher thread has terminated
// due to some error.
//
GMSEC_BOOL g_error = GMSEC_FALSE;

void myEventCallback(GMSEC_Connection conn, GMSEC_Status status, GMSEC_ConnectionEvent event)
{
	// Check event type and/or error status of the Status object.

	if (event == GMSEC_DISPATCHER_ERROR_EVENT)
	{
		// Auto-dispatcher error has occurred; we are done!
		GMSEC_ERROR("myEventCallback -- %s", statusGetReason(status));
		g_error = GMSEC_TRUE;
	}
	else if (statusHasError(status) == GMSEC_TRUE)
	{
		// Soft-error has occurred
		GMSEC_WARNING("myEventCallback -- %s", statusGetReason(status));
	}
	else
	{
		// Nominal event has occurred
		GMSEC_INFO("myEventCallback -- %s", statusGetReason(status));
	}
}

GMSEC_BOOL haveError()
{
	return g_error;
}


// Helper function
//
void checkStatus(const char* operation, GMSEC_Status status);


int main(int argc, char** argv)
{
	GMSEC_Status status = statusCreate();

	// Set up connection configuration options from the command line.
	GMSEC_Config config = configCreateWithArgs(argc, argv);

	// Initialize log level for output
	initializeLogLevel(config);

	// Deduce subscription topic for the example program
	const char* subscription_topic;
	GMSEC_I32 level = configGetIntegerValueOrDefault(config, "gmsec-schema-level", (GMSEC_I32) GMSEC_SCHEMA_LEVEL_0, NULL);

	switch (level)
	{
	case GMSEC_SCHEMA_LEVEL_0:
		subscription_topic = "C2MS.>";
		break;

	case GMSEC_SCHEMA_LEVEL_1:
		subscription_topic = "C2MS-PIPE.>";
		break;

	case GMSEC_SCHEMA_LEVEL_2:
	default:
		subscription_topic = "GMSEC.>";
		break;
	}

	// Create connection instance.
	GMSEC_Connection conn = connectionCreate(config, status);
	checkStatus("connectionCreate", status);

	// Output information regarding the API version and connection
	GMSEC_INFO(connectionGetAPIVersion());
	GMSEC_INFO("Middleware version = %s", connectionGetLibraryVersion(conn, NULL));

	// Register custom Event Callback
	connectionRegisterEventCallback(conn, GMSEC_ALL_EVENTS, &myEventCallback, status);
	checkStatus("connectionRegisterEventCallback", status);

	// Establish connection to the GMSEC Bus.
	connectionConnect(conn, status);
	checkStatus("connectionConnect", status);

	// Subscribe to receive message via a callback
	connectionSubscribeWithCallback(conn, subscription_topic, &myCallback, status);
	checkStatus("connectionSubscribe", status);
	GMSEC_INFO("Subscribed to topic: %s", subscription_topic);

	// Start the auto-dispatcher so we can receive messages via the callback
	connectionStartAutoDispatch(conn, status);
	checkStatus("connectionStartAutoDispatch", status);

	// Normally an application would have something important to do;
	// here we just idle until a message arrives. Should an error
	// occur, we bail.
	GMSEC_INFO("Waiting to receive a message...");
	while (gotMessage() == GMSEC_FALSE && haveError() == GMSEC_FALSE)
	{
		timeUtilMillisleep(1000);
	}

	// Disconnect from the GMSEC Bus, and terminate subscriptions.
	connectionDisconnect(conn, NULL);

	return 0;
}


void checkStatus(const char* operation, GMSEC_Status status)
{
	if (statusHasError(status) == GMSEC_TRUE)
	{
		GMSEC_ERROR("%s failed; reason: %s", operation, statusGet(status));
		exit(-1);
	}
}

File: examples/csharp/subscribe_async/subscribe_async.cs
/*
 * Copyright 2007-2025 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file subscribe_async.cs
 *
 * @brief This example program provides a basic demonstration of subscribing
 * to receive a message from the GMSEC Bus using an asynchronous callback.
 */

using GMSEC.API5;
using System;
using System.Text;


// Define custom callback to receive messages
//
class MyCallback : Callback
{
	private bool gotMsg;

	public MyCallback()
	{
		gotMsg = false;
	}

	public override void OnMessage(Connection conn, Message msg)
	{
		gotMsg = true;

		Log.Info("MyCallback.OnMessage -- \n" + msg.ToXML());

		// Acknowledge processing the message
		msg.Acknowledge();
	}

	public bool GotMessage()
	{
		return gotMsg;
	}
}


// Define custom callback to receive events
//
// Although not required, an event callback can assist with determining
// whether we have a connection to the GMSEC Bus, or in the case of this
// example program, whether the Auto-Dispatcher thread has terminated
// due to some error.
//
class MyEventCallback : EventCallback
{
	private bool error;

	public MyEventCallback()
	{
		error = false;
	}

	public override void OnEvent(Connection conn, Status status, Connection.Event theEvent)
	{
		// Check event type and/or error status of the Status object.

		if (theEvent == Connection.Event.DISPATCHER_ERROR_EVENT)
		{
			// Auto-dispatcher error has occurred; we are done!
			Log.Error("MyEventCallback.OnEvent -- " + status.GetReason());
			error = true;
		}
		else if (status.HasError())
		{
			// Soft-error has occurred
			Log.Warning("MyEventCallback.OnEvent -- " + status.GetReason());
		}
		else
		{
			// Nominal event has occurred
			Log.Info("MyEventCallback.OnEvent -- " + status.GetReason());
		}
	}

	public bool HaveError()
	{
		return error;
	}
}


public class subscribe_async
{
	public static int Main(string[] args)
	{
		// Set up connection configuration options from the command line.
		Config config = new Config(args);

		// Validate incoming messages
		config.AddValue("gmsec-msg-content-validate-recv", "true");

		// Initialize log level for output
		InitializeLogLevel(config);

		// Declared custom callback and custom event callback
		MyCallback      mcb = new MyCallback();
		MyEventCallback ecb = new MyEventCallback();

		// Deduce subscription topic for the example program
		string subscription_topic;
		int level = config.GetIntegerValue("gmsec-schema-level", (int) Specification.SchemaLevel.LEVEL_0);

		switch ((Specification.SchemaLevel) level)
		{
		case Specification.SchemaLevel.LEVEL_0:
			subscription_topic = "C2MS.>";
			break;

		case Specification.SchemaLevel.LEVEL_1:
			subscription_topic = "C2MS-PIPE.>";
			break;

		case Specification.SchemaLevel.LEVEL_2:
		default:
			subscription_topic = "GMSEC.>";
			break;
		}

		try
		{
			// Create connection instance.
			using (Connection conn = new Connection(config))
			{
				// Output information regarding the API version and connection
				Log.Info(Connection.GetAPIVersion());
				Log.Info("Middleware version = " + conn.GetLibraryVersion());

				// Register custom Event Callback
				conn.RegisterEventCallback(Connection.Event.ALL_EVENTS, ecb);

				// Establish connection to the GMSEC Bus.
				conn.Connect();

				// Subscribe to receive message via a callback
				conn.Subscribe(subscription_topic, mcb);
				Log.Info("Subscribe to topic: " + subscription_topic);

				// Start the auto-dispatcher so we can receive messages via the callback
				conn.StartAutoDispatch();

				// Normally an application would have something important to do;
				// here we just idle until a message arrives. Should an error
				// occur, we bail.
				Log.Info("Waiting to receive a message...");
				while (!mcb.GotMessage() && !ecb.HaveError())
				{
					TimeUtil.Millisleep(1000);
				}

				// Disconnect from the GMSEC Bus, and terminate subscriptions.
				conn.Disconnect();
			}
		}
		catch (GmsecException e)
		{
			// If here, the most likely case is the middleware could not be
			// deduced, or a failure occurred when attempting to connect to
			// the GMSEC Bus.
			Log.Error("GmsecException: " + e.ToString());
			return -1;
		}

		return 0;
	}


	private static void InitializeLogLevel(Config config)
	{
		LogLevel level = Log.FromString( config.GetValue("loglevel", "info") );
		Log.SetReportingLevel(level);
	}
}
